4-WebApp-your-API/4-1-MyOrg/TodoListService/Controllers/TodoListController.cs

```diff
@@ -23,70 +26,160 @@ public class TodoListController : Controller
+                TodoStore.Add(3, new Todo() { Id = 3, Owner = "Other User", Title = "Rent a car" });
+                TodoStore.Add(4, new Todo() { Id = 4, Owner = "Other User", Title = "Get vaccinated" });
             }
         }
 
-        // GET: api/values
         [HttpGet]
-        [RequiredScope(new string[] { "ToDoList.Read", "ToDoList.Write" })]
+        [RequiredScopeOrAppPermission(
+            AcceptedScope = new string[] { "ToDoList.Read", "ToDoList.ReadWrite" },
+            AcceptedAppPermission = new string[] { "ToDoList.Read.All", "ToDoList.ReadWrite.All" }
+            )]
         public IEnumerable<Todo> Get()
         {
-            string owner = User.Identity.Name;
-            return TodoStore.Values.Where(x => x.Owner == owner);
+            if (HasDelegatedPermissions(new string[] { "ToDoList.Read", "ToDoList.ReadWrite" }))
+            {
+                return TodoStore.Values.Where(x => x.Owner == _contextAccessor.HttpContext.User.GetObjectId());
+            }
+            else if (HasApplicationPermissions(new string[] { "ToDoList.Read.All", "ToDoList.ReadWrite.All" }))
+            {
+                return TodoStore.Values;
+            }
+
+            return null;
         }
 
-        // GET: api/values
         [HttpGet("{id}", Name = "Get")]
-        [RequiredScope(new string[] { "ToDoList.Read", "ToDoList.Write" })]
+        [RequiredScopeOrAppPermission(
+            AcceptedScope = new string[] { "ToDoList.Read", "ToDoList.ReadWrite" },
+            AcceptedAppPermission = new string[] { "ToDoList.Read.All", "ToDoList.ReadWrite.All" })]
         public Todo Get(int id)
         {
-            return TodoStore.Values.FirstOrDefault(t => t.Id == id);
+            //if it only has delegated permissions
+            //then it will be t.id==id && x.Owner == owner
+            //if it has app permissions the it will return  t.id==id
+
+            if (HasDelegatedPermissions(new string[] { "ToDoList.Read", "ToDoList.ReadWrite" }))
+            {
+                return TodoStore.Values.FirstOrDefault(t => t.Id == id && t.Owner == _contextAccessor.HttpContext.User.GetObjectId());
+            }
+            else if (HasApplicationPermissions(new string[] { "ToDoList.Read.All", "ToDoList.ReadWrite.All" }))
+            {
+                return TodoStore.Values.FirstOrDefault(t => t.Id == id);
+            }
+
+            return null;
         }
 
         [HttpDelete("{id}")]
-        [RequiredScope("ToDoList.Write")]
+        [RequiredScopeOrAppPermission(
+            AcceptedScope = new string[] { "ToDoList.ReadWrite" },
+            AcceptedAppPermission = new string[] { "ToDoList.ReadWrite.All" })]
         public void Delete(int id)
         {
-            TodoStore.Remove(id);
+            if (
+                (
+
+                HasDelegatedPermissions(new string[] { "ToDoList.ReadWrite" }) && TodoStore.Values.Any(x => x.Id == id && x.Owner == _contextAccessor.HttpContext.User.GetObjectId()))
```

Comment:
>_contextAccessor.HttpContext.User.GetObjectId() [](http://example.com/codeflow?start=133&length=47)

Again, you a repeating the same line of code over and over and not considering using a function !!!

Reply:
fixed